icu4c/source/test/intltest/measfmttest.cpp

```diff
@@ -3413,34 +3410,119 @@ void MeasureFormatTest::TestCompoundUnitOperations() {
     status.errIfFailureAndReset();
-    MeasureUnit powerN16a = MeasureUnit::forIdentifier("one-per-p16-kilometer", status);
+    MeasureUnit powerN16a = MeasureUnit::forIdentifier("per-p16-kilometer", status);
     status.expectErrorAndReset(U_ILLEGAL_ARGUMENT_ERROR);
     MeasureUnit powerN16b = powerN15.product(overQuarticKilometer1, status);
     status.expectErrorAndReset(U_ILLEGAL_ARGUMENT_ERROR);
 }
 
+void MeasureFormatTest::TestDimensionlessBehaviour() {
+    IcuTestErrorCode status(*this, "TestDimensionlessBehaviour");
+    MeasureUnit dimensionless;
+    MeasureUnit modified;
+
+    // At the time of writing, each of the seven groups below caused
+    // Parser::from("") to be called:
+
+    // splitToSingleUnits
+    int32_t count;
+    LocalArray<MeasureUnit> singles = dimensionless.splitToSingleUnits(count, status);
+    status.errIfFailureAndReset("dimensionless.splitToSingleUnits(...)");
+    assertEquals("no singles in dimensionless", 0, count);
+
+    // product(dimensionless)
+    MeasureUnit mile = MeasureUnit::getMile();
+    mile = mile.product(dimensionless, status);
+    status.errIfFailureAndReset("mile.product(dimensionless, ...)");
+    verifySingleUnit(mile, UMEASURE_SI_PREFIX_ONE, 1, "mile");
+
+    // dimensionless.getComplexity()
+    UMeasureUnitComplexity complexity = dimensionless.getComplexity(status);
+    status.errIfFailureAndReset("dimensionless.getComplexity(...)");
+    assertEquals("dimensionless complexity", UMEASURE_UNIT_SINGLE, complexity);
+
+    // dimensionless.getSIPrefix()
+    UMeasureSIPrefix siPrefix = dimensionless.getSIPrefix(status);
+    status.errIfFailureAndReset("dimensionless.getSIPrefix(...)");
+    assertEquals("dimensionless SIPrefix", UMEASURE_SI_PREFIX_ONE, siPrefix);
+
+    // dimensionless.withSIPrefix()
+    modified = dimensionless.withSIPrefix(UMEASURE_SI_PREFIX_KILO, status);
+    status.errIfFailureAndReset("dimensionless.withSIPrefix(...)");
+    singles = modified.splitToSingleUnits(count, status);
+    assertEquals("no singles in modified", 0, count);
+    siPrefix = modified.getSIPrefix(status);
+    status.errIfFailureAndReset("modified.getSIPrefix(...)");
+    assertEquals("modified SIPrefix", UMEASURE_SI_PREFIX_ONE, siPrefix);
+
+    // dimensionless.getDimensionality()
+    int32_t dimensionality = dimensionless.getDimensionality(status);
+    status.errIfFailureAndReset("dimensionless.getDimensionality(...)");
+    assertEquals("dimensionless dimensionality", 1, dimensionality);
+
+    // dimensionless.withDimensionality()
+    modified = dimensionless.withDimensionality(-1, status);
+    status.errIfFailureAndReset("dimensionless.withDimensionality(...)");
+    singles = modified.splitToSingleUnits(count, status);
+    assertEquals("no singles in modified", 0, count);
+    dimensionality = modified.getDimensionality(status);
+    status.errIfFailureAndReset("modified.getDimensionality(...)");
+    assertEquals("modified dimensionality", 1, dimensionality);
```

Comment:
Should dimensionality be 0 for the dimensionless unit?

Reply:
I understand why it returns 1, but I'm thinking that it's actually more correct to return 0.  It also gives code an easy way to detect the base dimensionless unit.

Spend 30 minutes seeing if you can make it work returning 0.  Maybe make SimpleUnitImpl default to 0 instead of 1, and set it to 1 at the top of the parse function after it's seen at least one token?  If you can get it to work, let's add that change.  If not, leave the value at 1 and perhaps file a follow-up ticket.

Reply:
This gets messy without setters and getters: I would love to maintain the invariant that if (index < 0), then (dimensionality == 0), and vice versa. However in the parsing, the dimensionality gets set first, before we know whether there will be an index to set. If the index does not get set, the dimensionality needs to get set back to 0 if we care about the invariant.

I already added isDimensionless() - I think we should just define "dimensionality" as meaningless (NaN!) when it's the dimensionless unit.

Reply:
My proposal (as per latest commit): getting explicit with dimensionality being as meaningless for dimensionless units as for compound units.